esx_service/cli/vmdkops_admin.py

```diff
@@ -723,38 +731,341 @@ def get_version():
+    headers = ['Uuid', 'Name', 'Description', 'Default_datastore', 'VM_list']
+    return headers
+
+def generate_vm_list(vms):
+    """ Generate a comma separated string with given list """
+    res = ""
+    for vm in vms:
+        # vm[1] is vm_name, vm has format (vm_uuid, vm_name)
+        res = res + vm[1]
+        res = res + ","
+        
+    if res:
+        res = res[:-1]
+
+    return res    
+
+def generate_tenant_ls_rows(tenant_list):
+    """ Generate output for tenant ls command """
+    rows = []
+    for tenant in tenant_list:
+        uuid = tenant.id
+        name = tenant.name
+        description = tenant.description
+        default_datastore = tenant.default_datastore
+        vm_list = generate_vm_list(tenant.vms)
+        rows.append([uuid, name, description, default_datastore, vm_list])
+    
+    return rows
 
-NOT_IMPLEMENTED = "Not implemented"
+def generate_tuple_from_vm_list(vm_list):
+    """ Generate a list of (vm_uuid, vm_name) pair """
+    vms = []
+    for vm_name in vm_list:
+        vm_uuid = vmdk_utils.get_vm_uuid_by_name(vm_name)
+        if not vm_uuid:
+            error_info =  "Cannot find vm_uuid for vm {0}, tenant_create failed".format(vm_name)
+            return error_info, None
+        vms.append((vm_uuid, vm_name))
+    
+    return None, vms
 
 def tenant_create(args):
-    print(NOT_IMPLEMENTED)
-
+    """ Handle tenant create command """
+    error_info, vms = generate_tuple_from_vm_list(args.vm_list)
+    if error_info:
+        return operation_fail(error_info)
+      
+    name = args.name 
+    description = "" 
+    default_datastore ="default_ds" 
+    default_privileges =  {}
+    privileges = []
+
+    auth.connect_auth_db()
+    error_info, tenant = auth._auth_mgr.create_tenant(name, description, default_datastore, default_privileges, vms, privileges)
+    
+    if error_info:
+        return operation_fail(error_info)
+ 
 def tenant_rm(args):
-    print(NOT_IMPLEMENTED)
+    """ Handle tenant rm command """
+    error_info, tenant = get_tenant_from_db(args.name)
+    if error_info:
+        return operation_fail(error_info)
+
+    remove_volumes = False
+    if args.remove_volumes:
+        if args.remove_volumes == 'True':
+            remove_volumes = True
+
+    error_info = auth._auth_mgr.remove_tenant(tenant.id, remove_volumes)
+    if error_info:
+        return operation_fail(error_info)
 
 def tenant_ls(args):
-    print(NOT_IMPLEMENTED)
+    """ Handle tenant ls command """
+    header = tenant_ls_headers()
+    auth.connect_auth_db()
+    error_info, tenant_list = auth._auth_mgr.list_tenants()
+    if error_info:
+        return operation_fail(error_info)
 
-def tenant_vm_add(args):
-    print(NOT_IMPLEMENTED)
+    rows = generate_tenant_ls_rows(tenant_list)
+    print(cli_table.create(header, rows)) 
 
+def tenant_vm_add(args):
+    """ Handle tenant vm add command """
+    error_info, vms = generate_tuple_from_vm_list(args.vm_list)
```

Comment:
tenant validation is done inside get_tenant_from_db. This function will do
1. connect the auth DB
2. get tenant with the given name from DB
If any error happened in above steps, an error will be return

If no errors when connect to the DB and reading from DB, but no tenant is found with the given name, then "tenant" will be None, 
and the following code will check whether the tenant exist or not.